Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+
+The `hostfxr` will also track active runtime in the process. Due to limitations (and to simplify implementation) this tracking will actually not look at the actual `coreclr` module (or try to communicate with the runtime in any way). Instead `hostfxr` itself will track the host context initialization. The first host context initialization in the process will represent the "loaded runtime". It is only possible to have one "loaded runtime" in the process. Any subsequent host context initialization will just "attach" to the "loaded runtime" instead of creating a new one.
+
+*Note: It is technically possible to initialize host context, and never call any "run" method on it, which would mean the runtime would not get loaded. Trying to initialize a second host context while the first one didn't start the runtime yet will for now be considered illegal. It's a pending issue to solve this problem, as COM activation will probably need to be able to safely initialize/run from multiple threads at a time.*
+
+``` C
+#define hostfxr_handle = void *;
+
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;
+    const char_t * dotnet_root;
+};
+```
+
+The `hostfxr_initialize_parameters` structure stores parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the hosting components. It's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
+
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - optional, if `argc` is `0` then `argv` is ignored.
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
```

Comment:
On first read through, the hostfxr_initialization_parameters feels more like an internal detail than something a developer should provide (eg. how would they know their dotnet_root at this point?).  What does having this provided by the host help with?

Reply:
It's there for two reasons:
* The current hosting API in `hostfxr` takes these parameters, so without them we could not emulate the existing scenarios over the new API. For example the muxer and apphost pass slightly different values for these. I do agree that for the simple cases it should not be necessary to specify these, that's why I mentioned we may make it optional. As described in the struct section:
  * `host_path` is passed to CoreCLR as the name of the "exe" - it's use by PAL to initialize some of its internal structures (like the process name, command line and so on). We might be able to deduce this via some OS-level APIs.
  * `dotnet_root` - where to look for shared frameworks and such. The muxer passes its own folder, `apphost` passes the parent/parent folder of the `hostfxr` it's using. We should be able to deduce it from the path of the `hostfxr` itself.
* Extensibility. The struct is versioned by the `size` field which will let us possibly add more value to it in the future without a need for a new API.

Reply:
I like the thought of having it optional.  I see the intent, so I would ask that the sample at the end of the document attempt to put real values in there... that would remove my concern.

Reply:
Tried to improve the sample to make it a bit more obvious what should go into these two fields. It's C++, so it's just pseudo code really - the real code is a bit more involved (and probably platform specific).